packages/fast-components/src/form-associated/index.ts

```diff
@@ -0,0 +1,263 @@
+import { attr, FastElement } from "@microsoft/fast-element";
+import { keyCodeEnter } from "@microsoft/fast-web-utilities";
+
+/**
+ * Disable member ordering to keep property callbacks
+ * grouped with property declaration
+ */
+/* tslint:disable:member-ordering */
+export abstract class FormAssociated<
+    T extends HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement
+> extends FastElement {
+    /**
+     * Must evaluate to true to enable elementInternals.
+     * Feature detects API support and resolve respectively
+     */
+    public static get formAssociated(): boolean {
+        return "ElementInternals" in window;
+    }
+
+    /**
+     * Returns the validty state of the element
+     */
+    public get validity(): ValidityState {
+        return FormAssociated.formAssociated
+            ? this.elementInternals.validity
+            : this.proxy.validity;
+    }
+
+    /**
+     * Retrieve a reference to the associated form.
+     * Returns null if not associated to any form.
+     */
+    public get form(): HTMLFormElement | null {
+        return FormAssociated.formAssociated
+            ? this.elementInternals.form
+            : this.proxy.form;
+    }
+
+    /**
+     * Retrieve the localized validation message,
+     * or custom validation message if set.
+     */
+    public get validationMessage(): string {
+        return FormAssociated.formAssociated
+            ? this.elementInternals.validationMessage
+            : this.proxy.validationMessage;
+    }
+
+    /**
+     * Whether the element will be validated when the
+     * form is submitted
+     */
+    public get willValidate(): boolean {
+        return FormAssociated.formAssociated
+            ? this.elementInternals.willValidate
+            : this.proxy.willValidate;
+    }
+
+    /**
+     * A reference to all associated label elements
+     */
+    public get labels(): Node[] {
+        if (FormAssociated.formAssociated) {
+            return Array.from(this.elementInternals.labels);
+        } else if (
+            this.proxy instanceof HTMLElement &&
+            this.proxy.ownerDocument &&
+            this.id
+        ) {
+            // Labels associated by wraping the element: <label><custom-element></custom-element></label>
+            const parentLabels = this.proxy.labels;
+            // Labels associated using the `for` attribute
+            const forLabels = Array.from(
+                this.proxy.ownerDocument.querySelectorAll(`[for='${this.id}']`)

```

Comment:
Rather than querying the ownerDocument, I think we need to call `getRootNode()` on `proxy`. This will either return the document or the nearest shadow dom root. This is important because if the component is used inside of another shadow dom, the query selector against document will not find the labels.

Reply:
good call - updated.